WalletWasabi.Fluent/ViewModels/Wallets/Send/ChangeAvoidanceSuggestionViewModel.cs

```diff
@@ -73,7 +76,7 @@ public partial class ChangeAvoidanceSuggestionViewModel : SuggestionViewModel
 	}
 
 	public static async Task<IEnumerable<ChangeAvoidanceSuggestionViewModel>> GenerateSuggestionsAsync(
```

Comment:
This method looks like a good start to me but ideally we can do much better. So now this method is awaited and we basically compute everything in serial fashion. So if `ChangelessTransactionCoinSelector.TryGetCoins` takes long time, we can't see any suggestion until it finishes (if it finishes at all!). 

I think it would be lovely to display `smallerSuggestion` immediately and show some "we are computing this suggestion" placeholder for BnB, once it would be done, it would appear too. If no BnB suggestion is found, then then one remove that suggestion or show a sad smiley ;-)

WDYT?

PS: I'm still thinking why `NormalizeSuggestions` is written as it is. I would expect that one can use `SortedDictionary<int, ChangeAvoidanceSuggestionViewModel>` (key being suggestion weight, priority.. the lower the number the higher priority) instead of list to avoid sorting it manually.

Reply:
> if `ChangelessTransactionCoinSelector.TryGetCoins` takes long time, we can't see any suggestion until it finishes (if it finishes at all!)

We only let it run for 15 seconds, then we cancel it. We can change that, if 15 sec feels too long.

> I think it would be lovely to display smallerSuggestion immediately and show some "we are computing this suggestion" placeholder for BnB, once it would be done it would appear too. If no BnB suggestion is found, then then one remove that suggestion or show a sad smiley ;-)

Yes, it would be really nice, but I'm afraid it's not that simple. What I want to have in this PR is a basic and working usage of the algorithm.
We can further improve it in the future, but IMHO it's time to put this coin selection task behind us. Pimping suggestions up is not high priority right now.

> PS: I'm still thinking why NormalizeSuggestions is written as it is. I would expect that one can use SortedDictionary<int, ChangeAvoidanceSuggestionViewModel> in Avalonia (key being suggestion weight, priority.. the lower the number the higher priority) instead of a list to avoid sorting it manually.

I really don't know why `NormalizeSuggestions` looks like as it does right now. If you believe it can be improved, please open a PR.  To be clear I absolutely agree with you. What's happening inside `NormalizeSuggestions` makes little sense to me (or I just don't see the logic).

Reply:
I attempted to do it here https://github.com/kiminuo/WalletWasabi/tree/feature/2022-01-23-wire-up-bnb-iasyncenumerator it's not polished (i.e. I'm not sure how to sort those suggestions by "extra cost")

Reply:
> I attempted to do it here https://github.com/kiminuo/WalletWasabi/tree/feature/2022-01-23-wire-up-bnb-iasyncenumerator ([compare against master](https://github.com/zkSNACKs/WalletWasabi/compare/master...kiminuo:feature/2022-01-23-wire-up-bnb-iasyncenumerator?expand=1)) it's not polished 

Thank you, i will test this.

> edit: But then one does not need to use list for suggestions anyway, I guess, because we know how many suggestions there are ... 3?

Well, if the target is low enough, and we can pay it with only one coin, then we don't have the `smallerSuggestion`. The `defaultSuggestion`, which is (I assume) the original Tx, and the `largerSuggestion` are always showed. Plus we have the new `bnbSuggestion` which is shown all the time AFAIK.

Reply:
Do we even need the `defaultSuggestion`? We don't show that, plus on fee adjustment we jump back to the original Tx.

Reply:
The point of the `NormalizeSuggestions` is not sorting. Before the BnB algorithm just a very initial (to be able to show something in UI) logic was used. And the goal was to generate one `Smaller` and one `Larger` transaction that doesn't generate change. Sometimes (depending on the input coins) it generated two smaller or two larger transactions, which makes no sense. `NormalizeSuggestions` tries to fix this case to only have one larger or one smaller (or both).

The UX was refactored several times but the logic for generating suggestions was never touched. Because it was waiting for the BnB algorithm). I did not follow the development of BnB so I am not really familiar with what it does and perhaps `NormalizeSuggestions` is obsoleted and not needed anymore.

> we know how many suggestions there are ... 3

Maximum 2. One larger and one smaller compared to the original.

## The Goal
The code regarding generating suggestions (I guess) is totally not optimal. This is simply because it was optimal at a point, but basically, everything is changed since then, but the logic for generating is not.

The goal is to generate two transactions, one that is smaller (smaller by payment perspective) than the original and one that is larger.
If one of them is not possible that is totally okay to have only one.
If either of them is not possible that is also okay.

The best is if the generated suggestions are as close to the payment amount as they could be.
If you already have one larger suggestion, but later you are able to generate a better (That is closer to the payment amount) just simply replace it. (same with the smaller one)

**Feel free to completely refactor everything**, because it is completely likely it should be refactored.

Reply:
How does one make a smaller suggestion? 
I thought `smallerSuggestion` would always be smaller than the original tx, because we take the smallest `SpentCoin` out.
But that is not the case, `smallerSuggestion` can give back both (`Less` and `More`), while `BnbSuggestion` will always give back `More`.
Can we modify the `smallerSuggestion`, so it always give back a `Less` suggestion?

Reply:
Smaller suggestion is implemented here https://github.com/zkSNACKs/WalletWasabi/blob/c4bc72f1ee3032b0e099b951063f415e41a5cab8/WalletWasabi.Fluent/ViewModels/Wallets/Send/ChangeAvoidanceSuggestionViewModel.cs#L91-L106 as you say. And if you can simulate that `smallerSuggestion` changes form `Less` to `More`, then most likely 

* `wallet.BuildTransaction` may behave differently, then what one expects
* `wallet.Synchronizer.UsdExchangeRate` is a dynamic value so it affects dollar value.

You can probably add some `Debug.WriteLine("");` helper lines in `ChangeAvoidanceSuggestionViewModel.cs` and watch it in MSVS Output window to see what is going on there.

Reply:
https://github.com/zkSNACKs/WalletWasabi/blob/c4bc72f1ee3032b0e099b951063f415e41a5cab8/WalletWasabi.Fluent/ViewModels/Wallets/Send/ChangeAvoidanceSuggestionViewModel.cs#L91-L102

This is already obsoleted. You can use `TransactionHelpers.BuildChangelessTransaction` just pass the coins to it.

> But that is not the case, smallerSuggestion can give back both (Less and More), while BnbSuggestion will always give back More.

I might remember wrong, it seems only the two less was the case. https://github.com/zkSNACKs/WalletWasabi/issues/6226#issuecomment-905857074

> The `defaultSuggestion`, which is (I assume) the original Tx is skipped by the `NormalizeSuggestions`

Yes, and with a tiny refactoring, the `defaultSuggestion` is not needed. It used to be next to the suggestions but this is changed since then. Just nobody took care of this code.

Since you can make sure the `smallerSuggestion` works like it never give larger (it seems it is ok) and the BnB always gives back a larger. `NormalizeSuggestions` can be removed completely and then `defaultSuggestion` can be removed as well.

Reply:
> This is already obsoleted. You can use `TransactionHelpers.BuildChangelessTransaction` just pass the coins to it.

Uff, yes, I forgot `smallerSuggestion` used the old one. 

> I might remember wrong, it seems only the two less was the case. [#6226 (comment)](https://github.com/zkSNACKs/WalletWasabi/issues/6226#issuecomment-905857074)

Hmm, for me it gave me back `More` suggestions almost 95% of the time. 
Anyway I modified the smallerTx to this: 
```
var smallerTransaction = await Task.Run(() => TransactionHelpers.BuildChangelessTransaction(
				wallet,
				destination,
				transactionInfo.UserLabels,
				transactionInfo.FeeRate,
				requestedTransaction
					.SpentCoins
					.OrderByDescending(x => x.Amount)
					.Skip(1),
				tryToSign: false
				));
```
so we take out (skip) the biggest `SpentCoin`. With this it gives back `Less` suggestions, but also can give back suggestions like: `~26 USD Less`. But with a sanity check we can choose not to show these.

> `NormalizeSuggestions` can be removed completely and then `defaultSuggestion` can be removed as well.

Done and done.

Reply:
> Hmm, for me it gave me back More suggestions almost 95% of the time.
Anyway I modified the smallerTx to this:

Even if your solution fixes the issue, there is something that doesn't work as we expect. So there might be a bug. It should be investigated.

Reply:
Yep, i'm looking into it rn.